lock-api/src/main/java/com/palantir/lock/watch/FailureCheckingLockWatchEventCache.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.watch;
+
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+
+import com.google.common.reflect.AbstractInvocationHandler;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+
+final class FailureCheckingLockWatchEventCache extends AbstractInvocationHandler {
+
+    static LockWatchEventCache newProxyInstance(LockWatchEventCache defaultCache) {
+        return (LockWatchEventCache) Proxy.newProxyInstance(
+                LockWatchEventCache.class.getClassLoader(),
+                new Class<?>[] {LockWatchEventCache.class},
+                new FailureCheckingLockWatchEventCache(defaultCache));
+    }
+
+    private final LockWatchEventCache defaultCache;
+    private final LockWatchEventCache noOpCache;
+
+    @GuardedBy("this")
+    private boolean hasFailed = false;
+
+    private FailureCheckingLockWatchEventCache(LockWatchEventCache defaultCache) {
+        this.defaultCache = defaultCache;
+        this.noOpCache = NoOpLockWatchEventCache.INSTANCE;
+    }
+
+    @Override
+    protected synchronized Object handleInvocation(Object proxy, Method method, Object[] args) throws Throwable {
+        if (hasFailed) {
+            return method.invoke(noOpCache, args);
+        } else {
+            hasFailed = true;
```

Comment:
This is quite clever, but kinda non-trivial to read and also not entirely correct. You are missing not failing on your expected exceptions!

I would rewrite it to:

```
private final LockWatchEventCache delegate;

private FailureCheckingLockWatchEventCache(LockWatchEventCache defaultCache) {
        this.delegate = defaultCache;
}
```

and then #handleInvocation:

```
try {
  return method.invoke(noOpCache, args);
} catch (Throwable e) {
  if (e instance of TransactionLockWatchFailedException) {
   throw e;
  }
  delegate = NoOpLockWatchEventCache.INSTANCE;
  // You still don't want to cause an outage, so should make the transaction retry.
  throw new TransactionLockWatchFailedException("Unexpected failure", e);
}
```
Or something to that effect.

Reply:
Done